### PR TITLE
Implemented possibility to add custom template engines

### PR DESCRIPTION
**Pull request by saturn4er**
Go templates are good, but jade/haml is better, I think. So here is resolution to add custom html compilers.

### CODE REVIEW

File: template.go [hunk 1]
```diff
@@ -31,11 +31,13 @@ import (
 
 var (
 	beegoTplFuncMap = make(template.FuncMap)
-	// beeTemplates caching map and supported template file extensions.
-	beeTemplates  = make(map[string]*template.Template)
+// beeTemplates caching map and supported template file extensions.
```
Comment (astaxie) [inline template.go]: it looks you didn't use gofmt with the file

**Comment by saturn4er** [comment]
@astaxie Fixed

File: template.go [hunk 2]
```diff
@@ -32,10 +32,12 @@ import (
 var (
 	beegoTplFuncMap = make(template.FuncMap)
 	// beeTemplates caching map and supported template file extensions.
-	beeTemplates  = make(map[string]*template.Template)
+	beeTemplates  = make(map[string]TemplateI)
 	templatesLock sync.RWMutex
 	// beeTemplateExt stores the template extension which will build
 	beeTemplateExt = []string{"tpl", "html"}
+	// BeeTemplatePreprocessors stores associations of extension -> preprocessor handler
+	BeeTemplateEngines = map[string]func(root, path string, funcs template.FuncMap) (TemplateI, error){}
```
Comment (astaxie) [inline template.go]: define `func(root, path string, funcs template.FuncMap)` as a function type?

Comment (astaxie) [inline template.go]: And I think `BeeTemplateEngines` don't need to export

**Comment by saturn4er** [comment]
@astaxie Ok?

File: template.go [hunk 3]
```diff
@@ -32,10 +32,12 @@ import (
 var (
 	beegoTplFuncMap = make(template.FuncMap)
 	// beeTemplates caching map and supported template file extensions.
-	beeTemplates  = make(map[string]*template.Template)
+	beeTemplates  = make(map[string]TemplateI)
 	templatesLock sync.RWMutex
 	// beeTemplateExt stores the template extension which will build
 	beeTemplateExt = []string{"tpl", "html"}
+	// BeeTemplatePreprocessors stores associations of extension -> preprocessor handler
```
Comment (astaxie) [inline template.go]: please also fix the comments name. It still use BeeTemplatePreprocessors

File: template.go [hunk 4]
```diff
@@ -156,11 +161,18 @@ func BuildTemplate(dir string, files ...string) error {
 		fmt.Printf("filepath.Walk() returned %v\n", err)
 		return err
 	}
+	buildAllFiles := len(files) == 0
```
Comment (JessonChan) [inline template.go]: we should avoid creating intermediate variable

File: template.go [hunk 5]
```diff
@@ -32,10 +32,12 @@ import (
 var (
 	beegoTplFuncMap = make(template.FuncMap)
 	// beeTemplates caching map and supported template file extensions.
-	beeTemplates  = make(map[string]*template.Template)
+	beeTemplates  = make(map[string]TemplateI)
```
Comment (JessonChan) [inline template.go]: interface  in golang is named with -er ,such as io.Writ**er**,io.Read**er**,http.Flush**er**

**Comment by JessonChan** [comment]
@saturn4er maybe you can git rebase to merge some commits into fewer commits(squash or fixup the go-formating commit)

File: template.go (hunk 4 above)
Comment (saturn4er) [inline template.go]: On each loops iterations Go will recalculate len(), isn't it?

Comment (JessonChan) [inline template.go]: It's not much of a problem.len() is fast enough as a  builtin function and this BuildTemplate function  will be called by the init method in product environment.

**Comment by astaxie** [comment]
@saturn4er could you also help to provider the example how to add jade/haml?

**Comment by saturn4er** [comment]
@astaxie https://gist.github.com/saturn4er/05e545984922784ace9b

File: template.go [hunk 6]
```diff
@@ -156,11 +162,18 @@ func BuildTemplate(dir string, files ...string) error {
...
 		return err
 	}
+	buildAllFiles := len(files) == 0
 	for _, v := range self.files {
 		for _, file := range v {
-			if len(files) == 0 || utils.InSlice(file, files) {
+			if buildAllFiles || utils.InSlice(file, files) {
 				templatesLock.Lock()
-				t, err := getTemplate(self.root, file, v...)
+				fileExt := filepath.Ext(file)[1:]
```
Comment (astaxie) [inline template.go]: if the template has no ext. it will crash here.

Comment (astaxie) [inline template.go]: @saturn4er could you try to fix this?

**Pull request merged**